### PR TITLE
Adding diagnostic messages when trimming winforms and wpf apps

### PR DESCRIPTION
**Pull request by LakshanF**
Trimming is dangerous for winforms and wpf and adding a warning as requested in issue #16716

In Winforms case, we expect to disable built-in COM support when trimming and hence fail. Added a msbuild property that can be used to enable COM if trimming is required. Wpf apps get a warning when trimming

### CODE REVIEW

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+            var publishCommand = new PublishCommand(testAsset);
+            publishCommand.Execute($"/p:PublishTrimmed=true")
+                .Should()
+                .Pass();
+
+        }
+
+        [WindowsOnlyRequiresMSBuildVersionFact("16.8.0")]
+        public void It_publishes_wpf_forms_app_with_warning()
```
Comment (sbomer) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs:69]:
```suggestion
        public void It_publishes_wpf_app_with_warning()
```

File: src/Tasks/Common/Resources/Strings.resx [hunk 2]
```diff
@@ -758,4 +758,12 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1163: Cannot emit symbols when publishing for .NET 5 with Crossgen2 using composite mode.</value>
     <comment>{StrBegin="NETSDK1163: "}</comment>
   </data>
+  <data name="TrimmingWindowsFormsRequiresCom" xml:space="preserve">
+    <value>NETSDK1164: Trimming Windows Forms is dangerous and requires 'BuiltInComSupport' to be set to 'true' </value>
```
Comment (sfoslund) [inline src/Tasks/Common/Resources/Strings.resx:762]: nit: We usually end our error messages with a period and no whitespace.

File: src/Assets/TestProjects/KitchenSink/TestApp/TestApp.csproj [hunk 3]
```diff
@@ -18,6 +18,7 @@
     <TieredCompilationQuickJit>true</TieredCompilationQuickJit>
     <TieredCompilationQuickJitForLoops>true</TieredCompilationQuickJitForLoops>
     <StartupHookSupport>false</StartupHookSupport>
+    <BuiltInComSupport>false</BuiltInComSupport>
```
Comment (sfoslund) [inline src/Assets/TestProjects/KitchenSink/TestApp/TestApp.csproj:21]: Is this change necessary? It doesn't look like you're using this project in any of your tests.

File: src/Tasks/Common/Resources/xlf/Strings.de.xlf [hunk 4]
```diff
@@ -4,259 +4,259 @@
     <body>
       <trans-unit id="AddResourceWithNonIntegerResource">
         <source>NETSDK1076: AddResource can only be used with integer resource types.</source>
-        <target state="translated">NETSDK1076: AddResource kann nur mit ganzzahligen Ressourcentypen verwendet werden.</target>
+        <target state="new">NETSDK1076: AddResource can only be used with integer resource types.</target>
```
Comment (sbomer) [inline src/Tasks/Common/Resources/xlf/Strings.de.xlf:7]: Make sure existing translations don't get overwritten :)

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            };
+            testProject.AdditionalProperties["UseWindowsForms"] = "true";
+            var testAsset = _testAssetsManager.CreateTestProject(testProject);
+            
+            var publishCommand = new PublishCommand(testAsset);
+            publishCommand.Execute($"/p:PublishTrimmed=true")
+                .Should()
+                .Fail()
+                .And
+                .HaveStdOutContaining(Strings.@TrimmingWindowsFormsRequiresCom);
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs:41]: We usually assert that the error code is present instead of the string itself (for example: `.HaveStdOutContaining("NETSDK1165")`

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 6]
```diff
@@ -106,7 +106,10 @@ Copyright (c) .NET Foundation. All rights reserved.
                  ResourceName="CannotHaveSingleFileWithoutRuntimeIdentifier" />
     <NETSdkError Condition="'$(PublishSingleFile)' == 'true' and '$(UseAppHost)' != 'true'"
                  ResourceName="CannotHaveSingleFileWithoutAppHost" />
-
+    <NETSdkError Condition="'$(UseWindowsForms)' == 'true' and '$(PublishTrimmed)' == 'true' and '$(BuiltInComSupport)' != 'true'"
```
Comment (sfoslund) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:109]: This logic doesn't error or warn when winforms apps that are publishing trimmed have `BuiltInComSupport` set to true, is that intended?

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        [WindowsOnlyRequiresMSBuildVersionFact("16.8.0")]
+        public void It_publishes_and_err_windows_forms_app()
+        {
+            var targetFramework = "net6.0-windows";
+            TestProject testProject = new TestProject()
+            {
+                Name = "WinformsFailTest",
+                ProjectSdk = "Microsoft.NET.Sdk",
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs:30]: nit: we don't need to specify the sdk when it's just the default, this line can be removed in all the tests.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+namespace Microsoft.NET.Publish.Tests
+{
+    public class GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps : SdkTest
+    {
+        public GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps(ITestOutputHelper log) : base(log)
+        {
+        }
+
+        [WindowsOnlyRequiresMSBuildVersionFact("16.8.0")]
+        public void It_publishes_and_err_windows_forms_app()
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs:24]: nit: It should fail here, right? So can we name it something like `It_errors_on_publishing_winforms_app`?

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace Microsoft.NET.Publish.Tests
+{
+    public class GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps : SdkTest
+    {
+        public GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps(ITestOutputHelper log) : base(log)
+        {
+        }
+
+        [WindowsOnlyRequiresMSBuildVersionFact("16.8.0")]
```
Comment (sfoslund) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishTrimmedWindowsFormsAndWPFApps.cs:23]: We can assume that CI machines all have MSBuild 16.8 or greater, so we can just use `[WindowsOnlyFact]` here.

**Review by sfoslund — COMMENTED** [review]
Lots of nits, but looks good overall!

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets (hunk 6 above)
Comment (sbomer) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:109]: What does it mean when `BuiltInComSupport` is the empty string (which is the default as written)? We should be consistent whether trimming or not.

I think it would be better to check `'$(BuiltInComSupport)' == 'false'`, and set it to false when trimming like we do for startup hooks:
https://github.com/dotnet/sdk/blob/9da281a438aa7cf9b80cd313f66917b87a4bc72e/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets#L27-L31

(The alternative would be to keep the `!= 'true'` and set `BuiltInComSupport` to `true` when not trimming.)

Comment (LakshanF) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:109]: Yes, it is intended. We generally do have warn for all trimmed apps at the end and winforms will have this warning as well. We expect that all WinForm apps will run through this error first and then need to explicitly take an action (set property to true) to go into trimming.

Comment (LakshanF) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:109]: This is a little tricky, since for BUILD we want all WinForm apps to get this warning. There is a parallel effort to add feature switch support for BuiltInComSupport but that might not get there by BUILD. As is, setting this property to 'true' for WinForm Apps will allow to go to trimming even though it is a no-op if the build-in COM support doesn't get in by BUILD.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets [hunk 10]
```diff
@@ -30,6 +30,16 @@ Copyright (c) .NET Foundation. All rights reserved.
     <StartupHookSupport>false</StartupHookSupport>
   </PropertyGroup>
 
+  <!-- We disable built-in Com support for trimmed apps here so that the feature
+       switch can flow to the runtimeconfig.json. Built-in Com is disabled
+       by default since they may require assemblies, types or members that
+       could be removed by the linker, causing a trimmed app to crash. -->
+  <PropertyGroup Condition="'$(BuiltInComSupport)' == '' And
+                            '$(PublishTrimmed)' == 'true' And
+                            $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))">
```
Comment (sfoslund) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:39]: So this property should be set for any framework with version greater than or equal to 6.0? Net framework and net standard as well as net core (in the event those ever exist :) )?

Comment (LakshanF) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:39]: Yes, this is the pattern that is followed by similar feature switches like StartupHookSupport that we added recently.

**Review by sfoslund — APPROVED** [review]

File: src/Tasks/Common/Resources/Strings.resx [hunk 11]
```diff
@@ -758,4 +758,12 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1163: Cannot emit symbols when publishing for .NET 5 with Crossgen2 using composite mode.</value>
     <comment>{StrBegin="NETSDK1163: "}</comment>
   </data>
-</root>
+  <data name="TrimmingWindowsFormsRequiresCom" xml:space="preserve">
+    <value>NETSDK1164: Trimming Windows Forms is dangerous and requires 'BuiltInComSupport' to be set to 'true'.</value>
```
Comment (sbomer) [inline src/Tasks/Common/Resources/Strings.resx:762]: I wonder if this should actually be two messages:
- winforms requires com, which is disabled when trimming (hard error to block the scenario initially)
- trimming with BuiltInComSupport true is dangerous (should warn even if the user set it explicitly because some features might not work when trimmed)

**Review by sbomer — APPROVED** [review]
Implementation LGTM, but I had a question about the warning UX. Thanks!

File: src/Tasks/Common/Resources/Strings.resx (hunk 11 above)
Comment (LakshanF) [reply src/Tasks/Common/Resources/Strings.resx:762]: Is it ok to have that warning for any windows app that sets this property to true and trim? If so, I prefer waiting for the feature switch support but let me know if we should handle it here.

Comment (sbomer) [reply src/Tasks/Common/Resources/Strings.resx:762]: I think handling it here would be most correct. Whether BuiltInComSupport is true or false, trimming may break apps that use COM, but at least we can explain this when BuiltInComSupport=false. I'm concerned that setting BuiltInComSupport=true makes errors go away and creates the expectation that things will work, when in reality trimming might still cause problems, as far as I understand. @vitek-karas do you have an opinion on it?

Comment (agocke) [inline src/Tasks/Common/Resources/Strings.resx:762]: This phrasing seems both too harsh and too mild. 😄 On the one hand, I think "dangerous" probably overstates things -- I didn't think it was any more dangerous than any other trim-incompatible APIs.

On the other hand, we need to be clear that the configuration here is unsupported and not recommended.

My proposal is the following

```suggestion
    <value>NETSDK1164: Windows Forms requires COM, which is not trim-compatible and is disabled by default when trimming. To enable COM and proceed, set `<BuiltInComSupport>true</BuiltInComSupport>`. This configuration is not supported or recommended.</value>
```

Comment (agocke) [reply src/Tasks/Common/Resources/Strings.resx:762]: But in this case it would be suppressed just because WinForms has `<SuppressTrimAnalysisWarnings>true` right? I don't see why we would make an exception just for COM warnings. It's not like they're any more addressable than the other ones.

File: src/Tasks/Common/Resources/Strings.resx [hunk 12]
```diff
@@ -758,4 +758,12 @@ The following are names of parameters or literal values and should not be transl
...
+    <comment>{StrBegin="NETSDK1164: "}</comment>
+  </data>
+  <data name="TrimmingWpfIsDangerous" xml:space="preserve">
+    <value>NETSDK1165: Trimming WPF is dangerous and might have issues at runtime.</value>
```
Comment (agocke) [inline src/Tasks/Common/Resources/Strings.resx:766]: Same feedback as for the above. I'd avoid the "dangerous" terminology and focus on trim-compatibility and that the combination is not supported/recommended.

File: src/Tasks/Common/Resources/Strings.resx (hunk 11 above)
Comment (vitek-karas) [reply src/Tasks/Common/Resources/Strings.resx:762]: I don't think the message should speak about COM. There are other things in WinForms which are problematic for trimming - not just COM (although COM is the biggest offender). I think the message should be something like:
```
Windows Forms is not supported with trimming enabled. Please go to https://aka.ms/... for more details.
```

I would introduce a new property which bypasses this warning/error and which gets us back to roughly 5.0 behavior.

If we stick to the current logic around COM:
Some WinForms apps will be broken because of `CustomResourceTypesSupport`, so we should then also ask people to enable that feature switch to get WinForms. And there might be others... we simply don't know because we didn't do the work for WinForms.

Comment (sbomer) [reply src/Tasks/Common/Resources/Strings.resx:762]: After some discussion yesterday I think it's ok to rely on the linker warnings (which will be suppressed, but that's a separate concern). My suggestion was based on my understanding that there were COM activation scenarios that the linker wouldn't be able to detect, but it sounds like that's not true (aside from some possible edge cases).

File: src/Tasks/Common/Resources/Strings.resx [hunk 13]
```diff
@@ -759,7 +759,7 @@ The following are names of parameters or literal values and should not be transl
     <comment>{StrBegin="NETSDK1163: "}</comment>
   </data>
   <data name="TrimmingWindowsFormsRequiresCom" xml:space="preserve">
-    <value>NETSDK1164: Trimming Windows Forms is dangerous and requires 'BuiltInComSupport' to be set to 'true'.</value>
+    <value>NETSDK1164: Windows Forms is not supported with trimming enabled. Please go to https://aka.ms/AAby5y8 for more details. To enable trimming and proceed, set `&lt;enableWindowsFormsTrimming&gt;true&lt;/enableWindowsFormsTrimming&gt;`. This configuration is not supported or recommended.</value>
```
Comment (vitek-karas) [inline src/Tasks/Common/Resources/Strings.resx:762]:
```suggestion
    <value>NETSDK1164: Windows Forms is not supported with trimming enabled. Please go to https://aka.ms/AAby5y8 for more details. To enable trimming and proceed, set `&lt;EnableWindowsFormsTrimming&gt;true&lt;/EnableWindowsFormsTrimming&gt;`. This configuration is not supported or recommended.</value>
```

File: src/Tasks/Common/Resources/Strings.resx [hunk 14]
```diff
@@ -758,4 +758,12 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1163: Cannot emit symbols when publishing for .NET 5 with Crossgen2 using composite mode.</value>
     <comment>{StrBegin="NETSDK1163: "}</comment>
   </data>
-</root>
+  <data name="TrimmingWindowsFormsRequiresCom" xml:space="preserve">
+    <value>NETSDK1164: Windows Forms is not supported with trimming enabled. Please go to https://aka.ms/AAby5y8 for more details. To enable trimming and proceed, set `&lt;enableWindowsFormsTrimming&gt;true&lt;/enableWindowsFormsTrimming&gt;`. This configuration is not supported or recommended.</value>
```
Comment (sbomer) [inline src/Tasks/Common/Resources/Strings.resx:762]: Is that link going to keep pointing to https://docs.microsoft.com/en-us/dotnet/core/deploying/trim-self-contained? If so, we already have a vanity link you can use here: https://aka.ms/dotnet-illink. If you plan on adding a new doc for this, it might be worth creating a separate vanity link like we did for the per-assembly warning: https://aka.ms/dotnet-illink/libraries (which will be updated once the new docs are available).

**Review by sbomer — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets [hunk 15]
```diff
@@ -30,14 +30,14 @@ Copyright (c) .NET Foundation. All rights reserved.
...
   </PropertyGroup>
 
-  <!-- We disable built-in Com support for trimmed apps here so that the feature
-       switch can flow to the runtimeconfig.json. Built-in Com is disabled
+  <!-- We disable Windows Forms Trimming support for trimmed apps here so that the feature
+       switch can flow to the runtimeconfig.json. enableWindowsFormsTrimming is disabled
        by default since they may require assemblies, types or members that
        could be removed by the linker, causing a trimmed app to crash. -->
-  <PropertyGroup Condition="'$(BuiltInComSupport)' == '' And
+  <PropertyGroup Condition="'$(enableWindowsFormsTrimming)' == '' And
```
Comment (vitek-karas) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets:37]: MSBuild properties are always PascalCased - so it should be `EnableWindowsFormsTrimming`.

**Comment by agocke** [comment]
@vitek-karas Would your intention bet that we turn the `EnableWinFormsTrimming` property to eventually set the appropriate feature flags? Is that meaningfully different from just turning on the feature flags by default in the winforms targets whenever trimming is enabled?

Should we just treat WinForms and WPF the same and give a blanket  "not trim compatible, not supported, not recommended" message?

**Review by agocke — APPROVED** [review]
LGTM, I presume we'll redirect the aka.ms links shortly

**Review by sbomer — APPROVED** [review]
LGTM, thanks!

**Review by vitek-karas — COMMENTED** [review]
Sorry I missed this - just one comment about casing in the message.

**Comment by devedse** [comment]
So for me trimming worked fine for my WPF .NET 5.0 app. How can I disable this error?

Currently the command shows the following:
```
C:\xGitPrivate\DeveImageOptimizerWPF>dotnet publish .\DeveImageOptimizerWPF\DeveImageOptimizerWPF.csproj -r win-x64 -c Release /p:PublishSingleFile=true /p:SelfContained=true /p:IncludeNativeLibrariesForSelfExtract=true /p:PublishTrimmed=true
Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\6.0.101\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(178,5): error NETSDK1168: WPF is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/wpf for more details. [C:\xGitPrivate\DeveImageOptimizerWPF\DeveImageOptimizerWPF\DeveImageOptimizerWPF.csproj]
```

See also this build:
https://github.com/devedse/DeveImageOptimizerWPF/runs/4907817278?check_suite_focus=true

**Comment by vitek-karas** [comment]
@devedse Trimming in .NET 6 has lot of changes compared to .NET 5. In .NET 5 trimming was in preview, in .NET 6 it's now fully supported for certain scenarios. WPF is not one of them. In fact we know for sure that a very simple WPF app is very broken in .NET 6 by default if trimming is allowed. That's why the error is there.

**Comment by devedse** [comment]
@vitek-karas , I managed to get it to work by changing my csproj file to the following:

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<OutputType>WinExe</OutputType>
		<TargetFramework>net6.0-windows</TargetFramework>
		<DisableWinExeOutputInference>true</DisableWinExeOutputInference>
		<ApplicationIcon>Icon.ico</ApplicationIcon>
		<Version>1.0.0.0</Version>
		<nullable>enable</nullable>
		<UseWPF>true</UseWPF>
		<StartupObject />
		<_SuppressWpfTrimError>true</_SuppressWpfTrimError>
		<BuiltInComInteropSupport>true</BuiltInComInteropSupport>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="DeveCoolLib" Version="1.0.509" />
		<PackageReference Include="DeveImageOptimizer" Version="1.0.589" />
		<PackageReference Include="Microsoft.Toolkit.Mvvm" Version="7.1.2" />
		<PackageReference Include="Ookii.Dialogs.Wpf" Version="5.0.1" />
		<PackageReference Include="PropertyChanged.Fody" Version="3.4.0" />
	</ItemGroup>

	<ItemGroup>
		<TrimmerRootAssembly Include="System.Private.CoreLib" />
	</ItemGroup>

</Project>
```

It took quite some time to figure it out. Especially since the `_SuppressWpfTrimError` isn't really documented. The untrimmed exe file is around 150mb while the trimmed one is 116mb. It's not as trimmed as in .NET 5 though, there is was around 100mb, but I'm still happy it works.

To figure out what TrimmerRootAssemblies I had to add I looked at the Event Viewer in combination with the decompiled method it errored on and then included all relevant DLLs. After that I removed them 1 by 1 and apparently the `System.Private.CoreLib` was the only one required.

**Comment by vitek-karas** [comment]
`_SuppressWpfTrimError` is not document on purpose - it's absolutely not supported.

I'm happy it works for you, that said, please test your app thoroughly, trimming can break lot of things in subtle ways.
